### PR TITLE
Heartbeat frame support for ballerina message broker

### LINKED ISSUE #218: Heartbeat frame support
### Description

Heartbeat frames tell the recipient that the sender is still alive. The rate and timing of heartbeat frames is negotiated during connection tuning.

#### Guidelines for implementers:

- Heartbeat frames MUST have a channel number of zero. A peer that receives an invalid heartbeat frame MUST raise a connection exception with reply code 501 (frame error).
- If the peer does not support heartbeating it MUST discard the heartbeat frame without signalling any error or fault.
- The client should start sending heartbeats after receiving a Connection.Tune method, and start monitoring heartbeats after receiving Connection.Open. The server should start sending and monitoring heartbeats after receiving Connection.Tune-Ok.
- The peer should make a best effort to send heartbeats at specified intervals. Heartbeats can be sent at any time. Any sent octet is a valid substitute for a heartbeat, thus heartbeats only have to be sent if no non-heartbeat AMQP traffic is sent for longer than one heartbeat interval. If a peer detects no incoming traffic (i.e. received octets) for two heartbeat intervals or longer, it should close the connection without following the Connection.Close/Close-Ok handshaking, and log an error.
- Heartbeating should continue until the connection socket is closed, including during and after
Connection.Close/Close-Ok handshaking.

### PR DESCRIPTION
**Pull request by hasithajayasundara**

### CODE REVIEW

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/frames/ConnectionTuneOk.java [hunk 1]
```diff
@@ -45,22 +49,31 @@ public ConnectionTuneOk(int channelMax, long frameMax, int heartbeat) {
...
     @Override
     public void handle(ChannelHandlerContext ctx, AmqpConnectionHandler connectionHandler) {
-        // TODO add tuning logic
+
+        if (heartbeat != 0) {
+            // Add an idle state handler to pipeline
+            ctx.pipeline().addFirst("idleStateHandler", new IdleStateHandler
+                    (0, 0, heartbeat)).addAfter("idleStateHandler", "idleConnectionListener", new
+                    IdleConnectionListener());
+        }
```
[the remark this answers is not in the document]
Comment (hasithajayasundara) [reply modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/frames/ConnectionTuneOk.java]: Done

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/frames/HeartbeatFrame.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ */
+public class HeartbeatFrame extends GeneralFrame {
+
+    public HeartbeatFrame() {
+
+        super((byte) 8, 0);
+    }
+
+    // Payload size - type, channel ,payload size ,end character
+    public long getPayloadSize() {
```
[the remark this answers is not in the document]
Comment (hasithajayasundara) [reply modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/frames/HeartbeatFrame.java]: Done

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/AmqpDecoder.java [hunk 3]
```diff
@@ -39,6 +40,7 @@
  * Netty based AMQP frame decoder.
  */
 public class AmqpDecoder extends ByteToMessageDecoder {
+
```
[the remark this answers is not in the document]
Comment (hasithajayasundara) [reply modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/AmqpDecoder.java]: Removed new lines from AmqpDecoder classs

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListener.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import io.ballerina.messaging.broker.amqp.codec.frames.HeartbeatFrame;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class handles idle connections.
+ */
+public class IdleConnectionListener extends ChannelDuplexHandler {
```
[the remark this answers is not in the document]
Comment (hasithajayasundara) [reply modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListener.java]: Done

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListener.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+     * If a peer detects no incoming traffic (i.e. received octets) for two heartbeat intervals or longer, it should
+     * close the connection without following the connection.
+     */
+    private static final int MAXIMUM_HEARTBEAT_COUNT = 3;
+    private int heartbeatCount;
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+
+        if ((++heartbeatCount) == MAXIMUM_HEARTBEAT_COUNT) {
```
[the remark this answers is not in the document]
Comment (hasithajayasundara) [reply modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListener.java]: Done

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListener.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Two heartbeats sent and timed out.Closing channel.");
```
[the remark this answers is not in the document]
Comment (hasithajayasundara) [reply modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListener.java]: Done

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListener.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            ctx.writeAndFlush(new HeartbeatFrame());
+        }
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+
+        IdleConnectionListener idleConnectionListener = (IdleConnectionListener) (ctx.pipeline().context
+                ("idleConnectionListener").handler());
+        idleConnectionListener.setHeartbeatCount(0);
```
[the remark this answers is not in the document]
Comment (hasithajayasundara) [reply modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListener.java]: Yeah we can.added changes accordingly

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListener.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        super.channelRead(ctx, msg);
+    }
+
+    /**
+     * Set heartbeatCount variable.
+     *
+     * @param value new value to set for heartbeatCount variable
+     */
+    public void setHeartbeatCount(int value) {
```
[the remark this answers is not in the document]
Comment (hasithajayasundara) [reply modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListener.java]: Done

File: modules/broker-amqp/src/test/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListenerTest.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.times;
+
+public class IdleConnectionListenerTest {
+
+    @Test
+    public void testIdleConnectionListener() {
```
[the remark this answers is not in the document]
Comment (hasithajayasundara) [reply modules/broker-amqp/src/test/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListenerTest.java]: renamed to testHeartbeatFrameSend

File: modules/broker-amqp/src/test/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListenerTest.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        Mockito.when(context.channel().close()).thenReturn(null);
+        for (int i = 0; i < 3; i++) {
+            idleConnectionListener.userEventTriggered(context, null);
+        }
+        Mockito.verify(context.channel(), times(1)).close();
+        Mockito.verify(context, atLeast(2)).writeAndFlush(heartbeatFrame.capture());
+    }
+
+    @Test
+    public void testIdleConnectionListenerChannelRead() throws Exception {
```
[the remark this answers is not in the document]
Comment (hasithajayasundara) [reply modules/broker-amqp/src/test/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListenerTest.java]: renamed to testHeartbeatCountReset..WDYT?

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListener.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+     * If a peer detects no incoming traffic (i.e. received octets) for two heartbeat intervals or longer, it should
+     * close the connection without following the connection.
+     */
+    private static final int MAXIMUM_HEARTBEAT_COUNT = 2;
+    private int heartbeatCount;
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+
+        if (heartbeatCount == MAXIMUM_HEARTBEAT_COUNT) {
```
Comment (Asitha) [inline modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/codec/handlers/IdleConnectionListener.java]: Shall we have this as heartbeatCount > MAXIMUM_HEARTBEAT_COUNT

**Pull request closed without merging**